// @devexpress/dx-react-chart-material-ui/dist/dx-react-chart-material-ui.cjs.js
/**
 * Bundle of @devexpress/dx-react-chart-material-ui
 * Generated: 2018-11-07
 * Version: 1.9.0
 * License: https://js.devexpress.com/Licensing
 */

'use strict';

if (typeof process === "undefined") { var process = { env: {} }; }

Object.defineProperty(exports, '__esModule', { value: true });

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var dxReactChart = require('@devexpress/dx-react-chart');
var colorManipulator = require('@material-ui/core/styles/colorManipulator');
var styles = require('@material-ui/core/styles');
var classNames = _interopDefault(require('classnames'));
var React = require('react');
var PropTypes = require('prop-types');
var List = _interopDefault(require('@material-ui/core/List'));
var ListItemText = _interopDefault(require('@material-ui/core/ListItemText'));
var ListItem = _interopDefault(require('@material-ui/core/ListItem'));
var Typography = _interopDefault(require('@material-ui/core/Typography'));

var classCallCheck = function (instance, Constructor) {
  if (!(instance instanceof Constructor)) {
    throw new TypeError("Cannot call a class as a function");
  }
};

var createClass = function () {
  function defineProperties(target, props) {
    for (var i = 0; i < props.length; i++) {
      var descriptor = props[i];
      descriptor.enumerable = descriptor.enumerable || false;
      descriptor.configurable = true;
      if ("value" in descriptor) descriptor.writable = true;
      Object.defineProperty(target, descriptor.key, descriptor);
    }
  }

  return function (Constructor, protoProps, staticProps) {
    if (protoProps) defineProperties(Constructor.prototype, protoProps);
    if (staticProps) defineProperties(Constructor, staticProps);
    return Constructor;
  };
}();

var _extends = Object.assign || function (target) {
  for (var i = 1; i < arguments.length; i++) {
    var source = arguments[i];

    for (var key in source) {
      if (Object.prototype.hasOwnProperty.call(source, key)) {
        target[key] = source[key];
      }
    }
  }

  return target;
};

var inherits = function (subClass, superClass) {
  if (typeof superClass !== "function" && superClass !== null) {
    throw new TypeError("Super expression must either be null or a function, not " + typeof superClass);
  }

  subClass.prototype = Object.create(superClass && superClass.prototype, {
    constructor: {
      value: subClass,
      enumerable: false,
      writable: true,
      configurable: true
    }
  });
  if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass;
};

var objectWithoutProperties = function (obj, keys) {
  var target = {};

  for (var i in obj) {
    if (keys.indexOf(i) >= 0) continue;
    if (!Object.prototype.hasOwnProperty.call(obj, i)) continue;
    target[i] = obj[i];
  }

  return target;
};

var possibleConstructorReturn = function (self, call) {
  if (!self) {
    throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
  }

  return call && (typeof call === "object" || typeof call === "function") ? call : self;
};

var getBorderColor = function getBorderColor(theme) {
  return theme.palette.type === 'light' ? colorManipulator.lighten(colorManipulator.fade(theme.palette.divider, 1), 0.88) : colorManipulator.darken(colorManipulator.fade(theme.palette.divider, 1), 0.68);
};

var withClassName = function withClassName() {
  for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
    args[_key] = arguments[_key];
  }

  var setClassName = function setClassName(_ref) {
    var className = _ref.className,
        classes = _ref.classes,
        restProps = objectWithoutProperties(_ref, ['className', 'classes']);
    return _extends({
      className: classNames(classes.root, className)
    }, restProps);
  };
  // TODO: First candidate to `compose` util?
  return function (Target) {
    return styles.withStyles.apply(undefined, args)(dxReactChart.withPatchedProps(setClassName)(Target));
  };
};

var styles$1 = function styles$$1(theme) {
  var _theme$typography = theme.typography,
      fontFamily = _theme$typography.fontFamily,
      fontSize = _theme$typography.fontSize,
      fontWeightLight = _theme$typography.fontWeightLight;

  return {
    root: {
      fontFamily: fontFamily,
      fontSize: fontSize,
      fontWeight: fontWeightLight,
      display: 'flex',
      flexDirection: 'column',
      padding: '10px'
    }
  };
};

var Root = withClassName(styles$1)(dxReactChart.Chart.Root);

var styles$2 = function styles$$1(theme) {
  var fontFamily = theme.typography.fontFamily;

  return {
    root: {
      fill: theme.palette.text.secondary,
      fontFamily: fontFamily,
      fontSize: 12,
      fontWeight: 400
    }
  };
};

var Label = withClassName(styles$2)(dxReactChart.Chart.Label);

var palette = ['#42A5F5', '#FF7043', '#9CCC65', '#FFCA28', '#26A69A', '#EC407A'];

var ChartWithPalette = function ChartWithPalette(_ref) {
  var children = _ref.children,
      props = objectWithoutProperties(_ref, ['children']);
  return React.createElement(
    dxReactChart.Chart,
    props,
    React.createElement(dxReactChart.Palette, { scheme: palette }),
    children
  );
};

ChartWithPalette.components = dxReactChart.Chart.components;

process.env.NODE_ENV !== "production" ? ChartWithPalette.propTypes = {
  children: PropTypes.node.isRequired
} : void 0;

var Chart = dxReactChart.withComponents({ Root: Root })(ChartWithPalette);
Chart.Label = Label;

var styles$3 = function styles$$1() {
  return {
    root: {
      paddingTop: 16,
      paddingBottom: 16
    }
  };
};

var Root$1 = withClassName(styles$3, { name: 'LegendRoot' })(List);

var styles$4 = function styles$$1() {
  return {
    root: {
      fontSize: 14,
      paddingLeft: 8,
      paddingRight: 8
    }
  };
};

var Label$1 = withClassName(styles$4, { name: 'LegendLabel' })(function (_ref) {
  var text = _ref.text,
      restProps = objectWithoutProperties(_ref, ['text']);
  return React.createElement(
    ListItemText,
    restProps,
    text
  );
});

var styles$5 = function styles$$1(theme) {
  return {
    root: {
      alignItems: 'center',
      paddingTop: theme.spacing.unit * 0.5,
      paddingBottom: theme.spacing.unit * 0.5
    }
  };
};

var Item = withClassName(styles$5, { name: 'LegendItem' })(ListItem);

var Legend = dxReactChart.withComponents({
  Root: Root$1, Item: Item, Label: Label$1
})(dxReactChart.Legend);

var Text = function (_React$PureComponent) {
  inherits(Text, _React$PureComponent);

  function Text() {
    classCallCheck(this, Text);
    return possibleConstructorReturn(this, (Text.__proto__ || Object.getPrototypeOf(Text)).apply(this, arguments));
  }

  createClass(Text, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          text = _props.text,
          restProps = objectWithoutProperties(_props, ['text']);

      return React.createElement(
        Typography,
        _extends({ component: 'h3', variant: 'h5' }, restProps),
        text
      );
    }
  }]);
  return Text;
}(React.PureComponent);

process.env.NODE_ENV !== "production" ? Text.propTypes = {
  text: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired
} : void 0;

var Title = dxReactChart.withComponents({ Text: Text })(dxReactChart.Title);

var styles$6 = function styles$$1(theme) {
  return {
    point: {
      fill: theme.palette.background.paper
    }
  };
};

var setClassName = function setClassName(_ref) {
  var classes = _ref.classes,
      restProps = objectWithoutProperties(_ref, ['classes']);

  if (restProps.state) {
    var className = restProps.className,
        rest = objectWithoutProperties(restProps, ['className']);

    return _extends({
      className: classNames(classes.point, className)
    }, rest);
  }
  return restProps;
};

var Point = styles.withStyles(styles$6)(dxReactChart.withPatchedProps(setClassName)(dxReactChart.ScatterSeries.Point));

var ScatterSeries = dxReactChart.withComponents({ Point: Point })(dxReactChart.ScatterSeries);

var styles$7 = function styles$$1(theme) {
  return {
    root: {
      stroke: getBorderColor(theme),
      shapeRendering: 'crispEdges'
    }
  };
};

var Line = withClassName(styles$7)(dxReactChart.Grid.Line);

var ValueGrid = dxReactChart.withComponents({ Line: Line })(dxReactChart.ValueGrid);

var ArgumentGrid = dxReactChart.withComponents({ Line: Line })(dxReactChart.ArgumentGrid);

var styles$8 = function styles$$1() {
  return {
    root: {
      shapeRendering: 'crispEdges'
    }
  };
};

var Root$2 = withClassName(styles$8)(dxReactChart.Axis.Root);

var styles$9 = function styles$$1(theme) {
  return {
    root: {
      stroke: getBorderColor(theme),
      shapeRendering: 'crispEdges'
    }
  };
};

var Tick = withClassName(styles$9)(dxReactChart.Axis.Tick);

var styles$a = function styles$$1(theme) {
  var fontFamily = theme.typography.fontFamily;

  return {
    root: {
      fill: theme.palette.text.secondary,
      fontFamily: fontFamily,
      fontSize: 12,
      fontWeight: 400,
      backgroundColor: theme.palette.background.paper
    }
  };
};

var Label$2 = withClassName(styles$a)(dxReactChart.Axis.Label);

var styles$b = function styles$$1(theme) {
  return {
    root: {
      stroke: getBorderColor(theme),
      shapeRendering: 'crispEdges'
    }
  };
};

var Line$1 = withClassName(styles$b)(dxReactChart.Axis.Line);

var ValueAxis = dxReactChart.withComponents({
  Root: Root$2, Tick: Tick, Label: Label$2, Line: Line$1
})(dxReactChart.ValueAxis);

var ArgumentAxis = dxReactChart.withComponents({
  Root: Root$2, Tick: Tick, Label: Label$2, Line: Line$1
})(dxReactChart.ArgumentAxis);

exports.PieSeries = dxReactChart.PieSeries;
exports.BarSeries = dxReactChart.BarSeries;
exports.LineSeries = dxReactChart.LineSeries;
exports.AreaSeries = dxReactChart.AreaSeries;
exports.SplineSeries = dxReactChart.SplineSeries;
exports.Chart = Chart;
exports.Legend = Legend;
exports.Title = Title;
exports.ScatterSeries = ScatterSeries;
exports.ValueGrid = ValueGrid;
exports.ArgumentGrid = ArgumentGrid;
exports.ValueAxis = ValueAxis;
exports.ArgumentAxis = ArgumentAxis;
